File: src/main/java/edu/ntu/mdpg7algo/models/Arena.java
package edu.ntu.mdpg7algo.models;

import lombok.Data;

import java.util.ArrayList;

/**
 * A virtual arena consisting of a virtual gird, obstacles, and a robot
 */
@Data
public class Arena {

    public static final int NUM_ROWS = 20;
    public static final int NUM_COLS = 20;
    public static final int CELL_WIDTH = 10;  // absolute cell width in centimeters
    public static final int CELL_HEIGHT = 10;  // absolute cell height in centimeters
    private ArrayList<Obstacle> obstacles;
    private Obstacle[][] obstaclesMatrix;  // stores the obstacle at corresponding grids [y][x]
    private Car car;

    public Arena() {
        obstacles = new ArrayList<>();
        obstaclesMatrix = new Obstacle[NUM_ROWS][NUM_COLS];
    }

    public void setObstacles(ArrayList<Obstacle> obstacles) {
        this.obstaclesMatrix = new Obstacle[NUM_ROWS][NUM_COLS];
        for (Obstacle obstacle : obstacles) {
            this.obstaclesMatrix[obstacle.getCellY()][obstacle.getCellX()] = obstacle;
        }
        this.obstacles = obstacles;
    }

    public void addObstacle(Obstacle obstacle) {
        obstacles.add(obstacle);
        obstaclesMatrix[obstacle.getCellY()][obstacle.getCellX()] = obstacle;
    }

    public void removeObstacle(int cellX, int cellY) {
        Obstacle obstacle = obstaclesMatrix[cellY][cellX];
        obstaclesMatrix[cellY][cellX] = null;
        if (obstacle != null) {
            obstacles.remove(obstacle);
        }
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < NUM_ROWS; i++) {
            sb.append("|");
            for (int j = 0; j < NUM_COLS; j++) {
                sb.append(obstaclesMatrix[i][j] == null ? " " : obstaclesMatrix[i][j].getFacing().toSymbol());
                sb.append("|");
            }
            sb.append("\n");
        }
        return sb.toString();
    }
}

File: python/mdpg7/models/arena.py
"""
Collection of obstacles
"""
from typing import List, Union, Any

from mdpg7.config.constants import ArenaConst
from mdpg7.models.obstacle import Obstacle


class Arena:
    
    def __init__(self):
        self.arena: List[List[Union[Obstacle, Any]]] = \
            [[None for _ in range(ArenaConst.NUM_COLS)] for _ in range(ArenaConst.NUM_ROWS)]
        self.obstacles: List[Obstacle] = list()
    
    def add_obstacle(self, obstacle: Obstacle):
        if self.arena[obstacle.cell_pos.cell_y][obstacle.cell_pos.cell_x] is None:
            self.arena[obstacle.cell_pos.cell_y][obstacle.cell_pos.cell_x] = obstacle
            self.obstacles.append(obstacle)
    
    def remove_obstacle(self, cell_x, cell_y):
        obstacle = self.arena[cell_y][cell_x]
        self.obstacles.remove(obstacle)
        self.arena[cell_y][cell_x] = None
